<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Get Started • predCrossVar</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Get Started">
<meta property="og:description" content="predCrossVar">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">predCrossVar</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/getting_started.html">Get Started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/getting_started.html">Get Started</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/wolfemd/predCrossVar/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="getting_started_files/header-attrs-2.5/header-attrs.js"></script><link href="getting_started_files/anchor-sections-1.0/anchor-sections.css" rel="stylesheet">
<script src="getting_started_files/anchor-sections-1.0/anchor-sections.js"></script><link href="getting_started_files/pagedtable-1.1/css/pagedtable.css" rel="stylesheet">
<script src="getting_started_files/pagedtable-1.1/js/pagedtable.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Get Started</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/wolfemd/predCrossVar/blob/master/vignettes/getting_started.Rmd"><code>vignettes/getting_started.Rmd</code></a></small>
      <div class="hidden name"><code>getting_started.Rmd</code></div>

    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/wolfemd/predCrossVar">predCrossVar</a></span><span class="op">)</span></code></pre></div>
<div id="intro" class="section level1">
<h1 class="hasAnchor">
<a href="#intro" class="anchor"></a>Intro</h1>
<p>The <strong>predCrossVar</strong> package contains a complete set of functions for the prediction of additive and dominance genetic variances and <em>co</em>-variances among full-siblings based on parents, following <span class="citation">(<a href="#ref-lehermeier2017" role="doc-biblioref">C. Lehermeier et al. 2017</a>; <a href="#ref-lehermeier2017a" role="doc-biblioref">Christina Lehermeier, Teyssèdre, and Schön 2017</a>)</span>. <strong>predCrossVar</strong> enables the prediction of genetic variance on a multi-trait selection index. Built for diploid organisms with phased, chromosome- or linkage-group ordered biallelic marker data, and a centimorgan-scale genetic map.</p>
<div id="predicting-genetic-variance" class="section level2">
<h2 class="hasAnchor">
<a href="#predicting-genetic-variance" class="anchor"></a>Predicting genetic variance</h2>
<p>Summarize formula for predicting genetic variance in crosses. Thanks to those whose work I have learned from and built upon.</p>
<ul>
<li><p>Lehermeier</p></li>
<li><p>Allier UCPC</p></li>
<li><p>Neyhart</p></li>
<li><p>Bijma et al.</p></li>
<li><p>Lynch &amp; Walsh 1998</p></li>
<li><p>Also <em>in silico</em> as in <strong>PopVar</strong></p></li>
<li><p>Wolfe et al. Genomic mate selection MS</p></li>
</ul>
</div>
</div>
<div id="notice-package-dev--status" class="section level1">
<h1 class="hasAnchor">
<a href="#notice-package-dev--status" class="anchor"></a>Notice: Package dev. status</h1>
<p>At this early development stage, package is unforgiving. There are not many arguments with defaults. Input formats are pretty specific, but well defined.</p>
<p>Especially in the case of many markers and using parallel processing, memory consumption can be potentially very large and needs to be considered. Hopefully there is sig. room for improvement in this area.</p>
</div>
<div id="example-data" class="section level1">
<h1 class="hasAnchor">
<a href="#example-data" class="anchor"></a>Example data</h1>
<p>I simulated 100 diploid individuals with 2 chromosomes, 10 QTL each, using the <a href="https://cran.r-project.org/web/packages/AlphaSimR/index.html"><strong>AlphaSimR</strong> library</a>. Two correlated traits with additive and dominance effects. Main use is to exemplify the input formats and usage of <strong>predCrossVar</strong> functions.</p>
<div id="inputs" class="section level2">
<h2 class="hasAnchor">
<a href="#inputs" class="anchor"></a>Inputs</h2>
<p>To predict cross variance, 3 types of input are required:</p>
<ol style="list-style-type: decimal">
<li>Phased Parental Haplotypes</li>
<li>Genetic Map (or pairwise-recombination frequencies)</li>
<li>Marker Effects</li>
</ol>
<p>We start with the included example simulation data. 100 individuals, 2 chromosomes with 10 QTL/chr.</p>
<div id="genetic-map" class="section level3">
<h3 class="hasAnchor">
<a href="#genetic-map" class="anchor"></a>Genetic map</h3>
<p>The usual starting place for recombination frequency is a centimorgan-scale genetic map.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="va">genmap</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/str.html">str</a></span><span class="op">(</span><span class="va">genmap</span><span class="op">)</span>
<span class="co">#&gt;  Named num [1:20] 0.0816 0.102 0.2653 0.3061 0.551 ...</span>
<span class="co">#&gt;  - attr(*, "names")= chr [1:20] "1_5" "1_6" "1_14" "1_16" ...</span></code></pre></div>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># utility function to compute matrix of pairwise recombination frequencies. </span>
<span class="va">recombFreqMat</span><span class="op">&lt;-</span><span class="fu"><a href="../reference/genmap2recombfreq.html">genmap2recombfreq</a></span><span class="op">(</span><span class="va">genmap</span>,nChr <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span><span class="op">(</span><span class="va">recombFreqMat</span><span class="op">)</span>
<span class="co">#&gt; [1] 20 20</span></code></pre></div>
<p>In practice, these matrices get big as they have dimension <span class="math inline">\(N_{snp} \times N_{snp}\)</span>. It is recommended to precompute and store the <code>recombFreqMat</code>.</p>
<p>In predicting genetic variance in the <span class="math inline">\(F_1\)</span> cross, the <code>recombFreqMat</code> the matrix <span class="math inline">\(1-(2 \times recombFreqMat)\)</span> is computed. To avoid having to do that separately for each family being predicted, I designed the prediction functions to require <span class="math inline">\(1-(2 \times recombFreqMat)\)</span> as input, to save a great deal of computation. <em>Therefore</em> pre-compute it and store <em>that</em> on disk. Probably this can be improved on!</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">recombFreqMat</span><span class="op">&lt;-</span><span class="fl">1</span><span class="op">-</span><span class="op">(</span><span class="fl">2</span><span class="op">*</span><span class="va">recombFreqMat</span><span class="op">)</span>
<span class="va">recombFreqMat</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span>
<span class="co">#&gt;            1_5       1_6      1_14      1_16      1_28</span>
<span class="co">#&gt; 1_5  1.0000000 0.9600054 0.6925693 0.6382791 0.3911064</span>
<span class="co">#&gt; 1_6  0.9600054 1.0000000 0.7214223 0.6648703 0.4074002</span>
<span class="co">#&gt; 1_14 0.6925693 0.7214223 1.0000000 0.9216104 0.5647181</span>
<span class="co">#&gt; 1_16 0.6382791 0.6648703 0.9216104 1.0000000 0.6127514</span>
<span class="co">#&gt; 1_28 0.3911064 0.4074002 0.5647181 0.6127514 1.0000000</span></code></pre></div>
</div>
<div id="phased-parental-haplotypes" class="section level3">
<h3 class="hasAnchor">
<a href="#phased-parental-haplotypes" class="anchor"></a>Phased parental haplotypes</h3>
<p>The example dataset <strong>haplos</strong> contains 2 chrom. with 10 loci each (columns) for 100 individuals (200 rows).</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="st">"haploMat"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span><span class="op">(</span><span class="va">haploMat</span><span class="op">)</span>
<span class="co">#&gt; [1] 200  20</span></code></pre></div>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">haploMat</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span>
<span class="co">#&gt;        1_5 1_6 1_14 1_16 1_28 1_31 1_33 1_35 1_40 1_41</span>
<span class="co">#&gt; 1_HapA   1   0    1    0    1    1    0    1    0    1</span>
<span class="co">#&gt; 1_HapB   1   0    1    0    0    1    0    0    1    0</span>
<span class="co">#&gt; 2_HapA   0   0    0    0    1    0    0    0    0    0</span>
<span class="co">#&gt; 2_HapB   1   0    0    1    1    0    1    1    1    0</span>
<span class="co">#&gt; 3_HapA   0   0    1    1    0    1    0    1    0    0</span></code></pre></div>
</div>
<div id="marker-effects" class="section level3">
<h3 class="hasAnchor">
<a href="#marker-effects" class="anchor"></a>Marker effects</h3>
<p>The example data come in the form of an list of row-matrices, which is useful for multivariate analyses we will get into later. There is also an example <strong>DomEffectsList</strong> available to load later.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="st">"AddEffectsList"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/str.html">str</a></span><span class="op">(</span><span class="va">AddEffectsList</span><span class="op">)</span>
<span class="co">#&gt; List of 2</span>
<span class="co">#&gt;  $ Trait1: num [1, 1:20] 0.218 0.03 0.232 0.343 -0.114 ...</span>
<span class="co">#&gt;   ..- attr(*, "dimnames")=List of 2</span>
<span class="co">#&gt;   .. ..$ : NULL</span>
<span class="co">#&gt;   .. ..$ : chr [1:20] "1_5" "1_6" "1_14" "1_16" ...</span>
<span class="co">#&gt;  $ Trait2: num [1, 1:20] -0.159 -0.297 -0.299 0.237 -0.716 ...</span>
<span class="co">#&gt;   ..- attr(*, "dimnames")=List of 2</span>
<span class="co">#&gt;   .. ..$ : NULL</span>
<span class="co">#&gt;   .. ..$ : chr [1:20] "1_5" "1_6" "1_14" "1_16" ...</span></code></pre></div>
</div>
</div>
</div>
<div id="single-trait-predictions" class="section level1">
<h1 class="hasAnchor">
<a href="#single-trait-predictions" class="anchor"></a>Single-trait predictions</h1>
<p>The first set of functions we will use are simplest. They only handle single traits. No covariance predictions.</p>
<div id="predict-one-cross" class="section level2">
<h2 class="hasAnchor">
<a href="#predict-one-cross" class="anchor"></a>Predict one cross</h2>
<p>To start, predict a single cross between parents <strong>“1”</strong> and <strong>“2”</strong></p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># For the univariate case, input effects should be a column matrix.</span>
<span class="va">addEffectsT1</span><span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/r/base/t.html">t</a></span><span class="op">(</span><span class="va">AddEffectsList</span><span class="op">$</span><span class="va">Trait1</span><span class="op">)</span>
<span class="va">addEffectsT2</span><span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/r/base/t.html">t</a></span><span class="op">(</span><span class="va">AddEffectsList</span><span class="op">$</span><span class="va">Trait2</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span><span class="op">(</span><span class="va">addEffectsT1</span><span class="op">)</span>
<span class="co">#&gt; [1] 20  1</span></code></pre></div>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Trait 1</span>
<span class="fu"><a href="../reference/predCrossVarA.html">predCrossVarA</a></span><span class="op">(</span>sireID <span class="op">=</span> <span class="st">"1"</span>, damID <span class="op">=</span> <span class="st">"2"</span>, 
              addEffects <span class="op">=</span> <span class="va">addEffectsT1</span>,
              haploMat <span class="op">=</span> <span class="va">haploMat</span>,
              recombFreqMat <span class="op">=</span> <span class="va">recombFreqMat</span><span class="op">)</span>
<span class="co">#&gt; [1] "Variances predicted for family: 1x2- took 0.001 mins"</span>
<span class="co">#&gt; # A tibble: 1 x 3</span>
<span class="co">#&gt;   varA[,1] segsnps    computetime</span>
<span class="co">#&gt;      &lt;dbl&gt; &lt;list&gt;           &lt;dbl&gt;</span>
<span class="co">#&gt; 1    0.557 &lt;chr [18]&gt;       0.035</span></code></pre></div>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Trait 2</span>
<span class="fu"><a href="../reference/predCrossVarA.html">predCrossVarA</a></span><span class="op">(</span>sireID <span class="op">=</span> <span class="st">"1"</span>, damID <span class="op">=</span> <span class="st">"2"</span>, 
              addEffects <span class="op">=</span> <span class="va">addEffectsT2</span>,
              haploMat <span class="op">=</span> <span class="va">haploMat</span>,
              recombFreqMat <span class="op">=</span> <span class="va">recombFreqMat</span><span class="op">)</span>
<span class="co">#&gt; [1] "Variances predicted for family: 1x2- took 0.001 mins"</span>
<span class="co">#&gt; # A tibble: 1 x 3</span>
<span class="co">#&gt;   varA[,1] segsnps    computetime</span>
<span class="co">#&gt;      &lt;dbl&gt; &lt;list&gt;           &lt;dbl&gt;</span>
<span class="co">#&gt; 1    0.502 &lt;chr [18]&gt;       0.047</span></code></pre></div>
</div>
<div id="predict-several-crosses" class="section level2">
<h2 class="hasAnchor">
<a href="#predict-several-crosses" class="anchor"></a>Predict several crosses</h2>
<p>Multiple families can be predicted, with an option for parallel processing across families with the <code><a href="../reference/runCrossVarPredsA.html">runCrossVarPredsA()</a></code> (additive-only) and <code><a href="../reference/runCrossVarPredsAD.html">runCrossVarPredsAD()</a></code> (additive-plus-dominance) functions.</p>
<p>For example, we will pick three parents of interest.</p>
<p>There is a helper function <code><a href="../reference/crosses2predict.html">crosses2predict()</a></code> to make a list of all pairwise matings given an input vecot of parent IDs. Includes self-crosses and the upper-triangle of a mating matrix (i.e. no reciprocal crosses).</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">ped2predict</span><span class="op">&lt;-</span><span class="fu"><a href="../reference/crosses2predict.html">crosses2predict</a></span><span class="op">(</span>parents <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"1"</span>,<span class="st">"2"</span>,<span class="st">"11"</span><span class="op">)</span><span class="op">)</span>
<span class="va">ped2predict</span>
<span class="co">#&gt; # A tibble: 6 x 2</span>
<span class="co">#&gt;   sireID damID</span>
<span class="co">#&gt;   &lt;chr&gt;  &lt;chr&gt;</span>
<span class="co">#&gt; 1 1      1    </span>
<span class="co">#&gt; 2 1      2    </span>
<span class="co">#&gt; 3 1      11   </span>
<span class="co">#&gt; 4 2      2    </span>
<span class="co">#&gt; 5 2      11   </span>
<span class="co">#&gt; 6 11     11</span></code></pre></div>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Trait 1 - serial (DEFAULT, ncores = 1)</span>
<span class="fu"><a href="../reference/runCrossVarPredsA.html">runCrossVarPredsA</a></span><span class="op">(</span>ped <span class="op">=</span> <span class="va">ped2predict</span>,
                  addEffects <span class="op">=</span> <span class="va">addEffectsT1</span>,
                  haploMat <span class="op">=</span> <span class="va">haploMat</span>,
                  recombFreqMat <span class="op">=</span> <span class="va">recombFreqMat</span><span class="op">)</span>
<span class="co">#&gt; Loading required package: furrr</span>
<span class="co">#&gt; Loading required package: future</span>
<span class="co">#&gt; [1] "Variances predicted for family: 1x1- took 0.001 mins"</span>
<span class="co">#&gt; [1] "Variances predicted for family: 1x2- took 0.001 mins"</span>
<span class="co">#&gt; [1] "Variances predicted for family: 1x11- took 0.001 mins"</span>
<span class="co">#&gt; [1] "Variances predicted for family: 2x2- took 0.001 mins"</span>
<span class="co">#&gt; [1] "Variances predicted for family: 2x11- took 0.001 mins"</span>
<span class="co">#&gt; [1] "Variances predicted for family: 11x11- took 0.001 mins"</span>
<span class="co">#&gt; [1] "Done predicting fam vars. Took 0.01 mins for 6 families"</span>
<span class="co">#&gt; # A tibble: 6 x 5</span>
<span class="co">#&gt;   sireID damID varA[,1] segsnps    computetime</span>
<span class="co">#&gt;   &lt;chr&gt;  &lt;chr&gt;    &lt;dbl&gt; &lt;list&gt;           &lt;dbl&gt;</span>
<span class="co">#&gt; 1 1      1        0.617 &lt;chr [9]&gt;       0.0580</span>
<span class="co">#&gt; 2 1      2        0.557 &lt;chr [18]&gt;      0.0500</span>
<span class="co">#&gt; 3 1      11       0.410 &lt;chr [18]&gt;      0.052 </span>
<span class="co">#&gt; 4 2      2        0.497 &lt;chr [12]&gt;      0.052 </span>
<span class="co">#&gt; 5 2      11       0.349 &lt;chr [19]&gt;      0.0550</span>
<span class="co">#&gt; 6 11     11       0.202 &lt;chr [11]&gt;      0.055</span></code></pre></div>
</div>
<div id="multi-core-across-families" class="section level2">
<h2 class="hasAnchor">
<a href="#multi-core-across-families" class="anchor"></a>Multi-core across families</h2>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Trait 1 - parallel (ncores = 6)</span>
<span class="fu"><a href="../reference/runCrossVarPredsA.html">runCrossVarPredsA</a></span><span class="op">(</span>ped <span class="op">=</span> <span class="va">ped2predict</span>,
                  addEffects <span class="op">=</span> <span class="va">addEffectsT1</span>,
                  haploMat <span class="op">=</span> <span class="va">haploMat</span>,
                  recombFreqMat <span class="op">=</span> <span class="va">recombFreqMat</span>,
                  ncores <span class="op">=</span> <span class="fl">6</span><span class="op">)</span>
<span class="co">#&gt; [1] "Variances predicted for family: 1x1- took 0.001 mins"</span>
<span class="co">#&gt; [1] "Variances predicted for family: 1x2- took 0.001 mins"</span>
<span class="co">#&gt; [1] "Variances predicted for family: 1x11- took 0.001 mins"</span>
<span class="co">#&gt; [1] "Variances predicted for family: 2x2- took 0.001 mins"</span>
<span class="co">#&gt; [1] "Variances predicted for family: 2x11- took 0.001 mins"</span>
<span class="co">#&gt; [1] "Variances predicted for family: 11x11- took 0.001 mins"</span>
<span class="co">#&gt; [1] "Done predicting fam vars. Took 0.02 mins for 6 families"</span>
<span class="co">#&gt; # A tibble: 6 x 5</span>
<span class="co">#&gt;   sireID damID varA[,1] segsnps    computetime</span>
<span class="co">#&gt;   &lt;chr&gt;  &lt;chr&gt;    &lt;dbl&gt; &lt;list&gt;           &lt;dbl&gt;</span>
<span class="co">#&gt; 1 1      1        0.617 &lt;chr [9]&gt;       0.03  </span>
<span class="co">#&gt; 2 1      2        0.557 &lt;chr [18]&gt;      0.03  </span>
<span class="co">#&gt; 3 1      11       0.410 &lt;chr [18]&gt;      0.031 </span>
<span class="co">#&gt; 4 2      2        0.497 &lt;chr [12]&gt;      0.0360</span>
<span class="co">#&gt; 5 2      11       0.349 &lt;chr [19]&gt;      0.032 </span>
<span class="co">#&gt; 6 11     11       0.202 &lt;chr [11]&gt;      0.032</span></code></pre></div>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Trait 2 - parallel (ncores = 6)</span>
<span class="fu"><a href="../reference/runCrossVarPredsA.html">runCrossVarPredsA</a></span><span class="op">(</span>ped <span class="op">=</span> <span class="va">ped2predict</span>,
                  addEffects <span class="op">=</span> <span class="va">addEffectsT2</span>,
                  haploMat <span class="op">=</span> <span class="va">haploMat</span>,
                  recombFreqMat <span class="op">=</span> <span class="va">recombFreqMat</span>,
                  ncores <span class="op">=</span> <span class="fl">6</span><span class="op">)</span>
<span class="co">#&gt; [1] "Variances predicted for family: 1x1- took 0 mins"</span>
<span class="co">#&gt; [1] "Variances predicted for family: 1x2- took 0 mins"</span>
<span class="co">#&gt; [1] "Variances predicted for family: 1x11- took 0.001 mins"</span>
<span class="co">#&gt; [1] "Variances predicted for family: 2x2- took 0.001 mins"</span>
<span class="co">#&gt; [1] "Variances predicted for family: 2x11- took 0.001 mins"</span>
<span class="co">#&gt; [1] "Variances predicted for family: 11x11- took 0.001 mins"</span>
<span class="co">#&gt; [1] "Done predicting fam vars. Took 0.02 mins for 6 families"</span>
<span class="co">#&gt; # A tibble: 6 x 5</span>
<span class="co">#&gt;   sireID damID varA[,1] segsnps    computetime</span>
<span class="co">#&gt;   &lt;chr&gt;  &lt;chr&gt;    &lt;dbl&gt; &lt;list&gt;           &lt;dbl&gt;</span>
<span class="co">#&gt; 1 1      1        0.637 &lt;chr [9]&gt;       0.0300</span>
<span class="co">#&gt; 2 1      2        0.502 &lt;chr [18]&gt;      0.028 </span>
<span class="co">#&gt; 3 1      11       0.505 &lt;chr [18]&gt;      0.0330</span>
<span class="co">#&gt; 4 2      2        0.367 &lt;chr [12]&gt;      0.033 </span>
<span class="co">#&gt; 5 2      11       0.370 &lt;chr [19]&gt;      0.042 </span>
<span class="co">#&gt; 6 11     11       0.373 &lt;chr [11]&gt;      0.0350</span></code></pre></div>
</div>
<div id="additive-plus-dominance" class="section level2">
<h2 class="hasAnchor">
<a href="#additive-plus-dominance" class="anchor"></a>Additive-plus-dominance</h2>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="st">"DomEffectsList"</span><span class="op">)</span>
<span class="va">domEffectsT1</span><span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/r/base/t.html">t</a></span><span class="op">(</span><span class="va">DomEffectsList</span><span class="op">$</span><span class="va">Trait1</span><span class="op">)</span>
<span class="va">domEffectsT2</span><span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/r/base/t.html">t</a></span><span class="op">(</span><span class="va">DomEffectsList</span><span class="op">$</span><span class="va">Trait2</span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Trait 1</span>
<span class="fu"><a href="../reference/runCrossVarPredsAD.html">runCrossVarPredsAD</a></span><span class="op">(</span>ped <span class="op">=</span> <span class="va">ped2predict</span>,
                  addEffects <span class="op">=</span> <span class="va">addEffectsT1</span>,
                  domEffects <span class="op">=</span> <span class="va">domEffectsT1</span>,
                  haploMat <span class="op">=</span> <span class="va">haploMat</span>,
                  recombFreqMat <span class="op">=</span> <span class="va">recombFreqMat</span>,
                  ncores <span class="op">=</span> <span class="fl">6</span><span class="op">)</span>
<span class="co">#&gt; [1] "Variances predicted for family: 1x1- took 0 mins"</span>
<span class="co">#&gt; [1] "Variances predicted for family: 1x2- took 0.001 mins"</span>
<span class="co">#&gt; [1] "Variances predicted for family: 1x11- took 0 mins"</span>
<span class="co">#&gt; [1] "Variances predicted for family: 2x2- took 0.001 mins"</span>
<span class="co">#&gt; [1] "Variances predicted for family: 2x11- took 0.001 mins"</span>
<span class="co">#&gt; [1] "Variances predicted for family: 11x11- took 0 mins"</span>
<span class="co">#&gt; [1] "Done predicting fam vars. Took 0.02 mins for 6 families"</span>
<span class="co">#&gt; # A tibble: 6 x 6</span>
<span class="co">#&gt;   sireID damID varA[,1] varD[,1] segsnps    computetime</span>
<span class="co">#&gt;   &lt;chr&gt;  &lt;chr&gt;    &lt;dbl&gt;    &lt;dbl&gt; &lt;list&gt;           &lt;dbl&gt;</span>
<span class="co">#&gt; 1 1      1        0.617    0.349 &lt;chr [9]&gt;       0.029 </span>
<span class="co">#&gt; 2 1      2        0.557    0.363 &lt;chr [18]&gt;      0.03  </span>
<span class="co">#&gt; 3 1      11       0.410    0.266 &lt;chr [18]&gt;      0.029 </span>
<span class="co">#&gt; 4 2      2        0.497    0.497 &lt;chr [12]&gt;      0.03  </span>
<span class="co">#&gt; 5 2      11       0.349    0.355 &lt;chr [19]&gt;      0.034 </span>
<span class="co">#&gt; 6 11     11       0.202    0.349 &lt;chr [11]&gt;      0.0270</span></code></pre></div>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Trait 2</span>
<span class="fu"><a href="../reference/runCrossVarPredsAD.html">runCrossVarPredsAD</a></span><span class="op">(</span>ped <span class="op">=</span> <span class="va">ped2predict</span>,
                  addEffects <span class="op">=</span> <span class="va">addEffectsT1</span>,
                  domEffects <span class="op">=</span> <span class="va">domEffectsT1</span>,
                  haploMat <span class="op">=</span> <span class="va">haploMat</span>,
                  recombFreqMat <span class="op">=</span> <span class="va">recombFreqMat</span>,
                  ncores <span class="op">=</span> <span class="fl">6</span><span class="op">)</span>
<span class="co">#&gt; [1] "Variances predicted for family: 1x1- took 0 mins"</span>
<span class="co">#&gt; [1] "Variances predicted for family: 1x2- took 0 mins"</span>
<span class="co">#&gt; [1] "Variances predicted for family: 1x11- took 0.001 mins"</span>
<span class="co">#&gt; [1] "Variances predicted for family: 2x2- took 0.001 mins"</span>
<span class="co">#&gt; [1] "Variances predicted for family: 2x11- took 0.001 mins"</span>
<span class="co">#&gt; [1] "Variances predicted for family: 11x11- took 0 mins"</span>
<span class="co">#&gt; [1] "Done predicting fam vars. Took 0.02 mins for 6 families"</span>
<span class="co">#&gt; # A tibble: 6 x 6</span>
<span class="co">#&gt;   sireID damID varA[,1] varD[,1] segsnps    computetime</span>
<span class="co">#&gt;   &lt;chr&gt;  &lt;chr&gt;    &lt;dbl&gt;    &lt;dbl&gt; &lt;list&gt;           &lt;dbl&gt;</span>
<span class="co">#&gt; 1 1      1        0.617    0.349 &lt;chr [9]&gt;       0.0280</span>
<span class="co">#&gt; 2 1      2        0.557    0.363 &lt;chr [18]&gt;      0.029 </span>
<span class="co">#&gt; 3 1      11       0.410    0.266 &lt;chr [18]&gt;      0.03  </span>
<span class="co">#&gt; 4 2      2        0.497    0.497 &lt;chr [12]&gt;      0.0310</span>
<span class="co">#&gt; 5 2      11       0.349    0.355 &lt;chr [19]&gt;      0.03  </span>
<span class="co">#&gt; 6 11     11       0.202    0.349 &lt;chr [11]&gt;      0.0290</span></code></pre></div>
</div>
</div>
<div id="multiple-trait-predictions" class="section level1">
<h1 class="hasAnchor">
<a href="#multiple-trait-predictions" class="anchor"></a>Multiple-trait predictions</h1>
<p>(including covariances)</p>
<p><code><a href="../reference/predOneCrossVarA.html">predOneCrossVarA()</a></code> -&gt; <code><a href="../reference/predCrossVarsA.html">predCrossVarsA()</a></code> -&gt; <code><a href="../reference/runMtCrossVarPredsA.html">runMtCrossVarPredsA()</a></code></p>
<p><code><a href="../reference/runMtCrossVarPredsA.html">runMtCrossVarPredsA()</a></code> and <code><a href="../reference/runMtCrossVarPredsAD.html">runMtCrossVarPredsAD()</a></code> should supersede in function all the others.</p>
<p><strong>NOTES:</strong></p>
<ul>
<li>
<p>Arguments names are pretty opinionated and long, although not without good intention. Might be worth simplifying.</p>
<ul>
<li>e.g. <code><a href="../reference/predOneCrossVarA.html">predOneCrossVarA()</a></code>: <code>postMeanAlleleSubEffects=</code> could just be <code>addEffectsLists=</code>
</li>
</ul>
</li>
</ul>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">mtpred</span><span class="op">&lt;-</span><span class="fu"><a href="../reference/runMtCrossVarPredsAD.html">runMtCrossVarPredsAD</a></span><span class="op">(</span>CrossesToPredict <span class="op">=</span> <span class="va">ped2predict</span>,
                             AddEffectList <span class="op">=</span> <span class="va">AddEffectsList</span>,
                             DomEffectList <span class="op">=</span> <span class="va">DomEffectsList</span>,
                             haploMat <span class="op">=</span> <span class="va">haploMat</span>,
                             recombFreqMat <span class="op">=</span> <span class="va">recombFreqMat</span>, 
                             ncores <span class="op">=</span> <span class="fl">6</span><span class="op">)</span>
<span class="co">#&gt; [1] "Variances predicted for family: 1x1- took 0 mins"</span>
<span class="co">#&gt; [1] "Variances predicted for family: 1x2- took 0.001 mins"</span>
<span class="co">#&gt; [1] "Variances predicted for family: 1x11- took 0.001 mins"</span>
<span class="co">#&gt; [1] "Variances predicted for family: 2x2- took 0.001 mins"</span>
<span class="co">#&gt; [1] "Variances predicted for family: 2x11- took 0.001 mins"</span>
<span class="co">#&gt; [1] "Variances predicted for family: 11x11- took 0.001 mins"</span>
<span class="co">#&gt; [1] "Done predicting fam vars. Took 0.05 mins for 6 families"</span>
<span class="co">#&gt; [1] "Variances predicted for family: 1x1- took 0.001 mins"</span>
<span class="co">#&gt; [1] "Variances predicted for family: 1x2- took 0.001 mins"</span>
<span class="co">#&gt; [1] "Variances predicted for family: 1x11- took 0 mins"</span>
<span class="co">#&gt; [1] "Variances predicted for family: 2x2- took 0.001 mins"</span>
<span class="co">#&gt; [1] "Variances predicted for family: 2x11- took 0.001 mins"</span>
<span class="co">#&gt; [1] "Variances predicted for family: 11x11- took 0.001 mins"</span>
<span class="co">#&gt; [1] "Done predicting fam vars. Took 0.05 mins for 6 families"</span>
<span class="co">#&gt; [1] "Variances predicted for family: 1x1- took 0.001 mins"</span>
<span class="co">#&gt; [1] "Variances predicted for family: 1x2- took 0.001 mins"</span>
<span class="co">#&gt; [1] "Variances predicted for family: 1x11- took 0.001 mins"</span>
<span class="co">#&gt; [1] "Variances predicted for family: 2x2- took 0.001 mins"</span>
<span class="co">#&gt; [1] "Variances predicted for family: 2x11- took 0.001 mins"</span>
<span class="co">#&gt; [1] "Variances predicted for family: 11x11- took 0 mins"</span>
<span class="co">#&gt; [1] "Done predicting fam vars. Took 0.05 mins for 6 families"</span></code></pre></div>
<p>The results are a nested list structure. Let’s unpack them…</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://tidyverse.tidyverse.org">tidyverse</a></span><span class="op">)</span>; <span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://magrittr.tidyverse.org">magrittr</a></span><span class="op">)</span>
<span class="co">#&gt; ── Attaching packages ─────────────────────────────────────── tidyverse 1.3.0 ──</span>
<span class="co">#&gt; ✓ ggplot2 3.3.2     ✓ dplyr   1.0.2</span>
<span class="co">#&gt; ✓ tibble  3.0.4     ✓ stringr 1.4.0</span>
<span class="co">#&gt; ✓ tidyr   1.1.2     ✓ forcats 0.5.0</span>
<span class="co">#&gt; ✓ readr   1.4.0</span>
<span class="co">#&gt; ── Conflicts ────────────────────────────────────────── tidyverse_conflicts() ──</span>
<span class="co">#&gt; x dplyr::filter() masks stats::filter()</span>
<span class="co">#&gt; x dplyr::lag()    masks stats::lag()</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Attaching package: 'magrittr'</span>
<span class="co">#&gt; The following object is masked from 'package:tidyr':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     extract</span>
<span class="co">#&gt; The following object is masked from 'package:purrr':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     set_names</span></code></pre></div>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">mtpred</span>
<span class="co">#&gt; $varcovars</span>
<span class="co">#&gt; # A tibble: 3 x 3</span>
<span class="co">#&gt;   Trait1 Trait2 varcomps        </span>
<span class="co">#&gt;   &lt;chr&gt;  &lt;chr&gt;  &lt;list&gt;          </span>
<span class="co">#&gt; 1 Trait1 Trait1 &lt;named list [2]&gt;</span>
<span class="co">#&gt; 2 Trait2 Trait2 &lt;named list [2]&gt;</span>
<span class="co">#&gt; 3 Trait1 Trait2 &lt;named list [2]&gt;</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $totcomputetime</span>
<span class="co">#&gt; elapsed </span>
<span class="co">#&gt;   9.268</span></code></pre></div>
<p>The list element <strong>“varcovars”</strong></p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">mtpred</span><span class="op">$</span><span class="va">varcovars</span> <span class="op">%&gt;%</span> 
  <span class="fu">tidyr</span><span class="fu">::</span><span class="fu"><a href="https://tidyr.tidyverse.org/reference/hoist.html">unnest_wider</a></span><span class="op">(</span><span class="va">varcomps</span><span class="op">)</span>
<span class="co">#&gt; # A tibble: 3 x 4</span>
<span class="co">#&gt;   Trait1 Trait2 predictedfamvars totcomputetime</span>
<span class="co">#&gt;   &lt;chr&gt;  &lt;chr&gt;  &lt;list&gt;                    &lt;dbl&gt;</span>
<span class="co">#&gt; 1 Trait1 Trait1 &lt;tibble [6 × 3]&gt;           2.90</span>
<span class="co">#&gt; 2 Trait2 Trait2 &lt;tibble [6 × 3]&gt;           3.07</span>
<span class="co">#&gt; 3 Trait1 Trait2 &lt;tibble [6 × 3]&gt;           3.29</span></code></pre></div>
<p>Unpack <strong>varcovars</strong> to get at the variance and covariance predictions for additive <em>and</em> dominances.</p>
<p>Variances are in column <strong>VPM</strong>. Output is in long-form.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">mtpred</span><span class="op">$</span><span class="va">varcovars</span> <span class="op">%&gt;%</span> 
  <span class="fu">tidyr</span><span class="fu">::</span><span class="fu"><a href="https://tidyr.tidyverse.org/reference/hoist.html">unnest_wider</a></span><span class="op">(</span><span class="va">varcomps</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span><span class="op">(</span><span class="op">-</span><span class="va">totcomputetime</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu">tidyr</span><span class="fu">::</span><span class="fu"><a href="https://tidyr.tidyverse.org/reference/nest.html">unnest</a></span><span class="op">(</span><span class="va">predictedfamvars</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu">tidyr</span><span class="fu">::</span><span class="fu"><a href="https://tidyr.tidyverse.org/reference/nest.html">unnest</a></span><span class="op">(</span><span class="va">predVars</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="va">head</span>
<span class="co">#&gt; # A tibble: 6 x 9</span>
<span class="co">#&gt;   Trait1 Trait2 sireID damID VarComp   VPM PMV   Nsegsnps totcomputetime</span>
<span class="co">#&gt;   &lt;chr&gt;  &lt;chr&gt;  &lt;chr&gt;  &lt;chr&gt; &lt;chr&gt;   &lt;dbl&gt; &lt;lgl&gt;    &lt;int&gt;          &lt;dbl&gt;</span>
<span class="co">#&gt; 1 Trait1 Trait1 1      1     VarA    0.617 NA           9          0.028</span>
<span class="co">#&gt; 2 Trait1 Trait1 1      1     VarD    0.349 NA          NA         NA    </span>
<span class="co">#&gt; 3 Trait1 Trait1 1      2     VarA    0.557 NA          18          0.032</span>
<span class="co">#&gt; 4 Trait1 Trait1 1      2     VarD    0.363 NA          NA         NA    </span>
<span class="co">#&gt; 5 Trait1 Trait1 1      11    VarA    0.410 NA          18          0.033</span>
<span class="co">#&gt; 6 Trait1 Trait1 1      11    VarD    0.266 NA          NA         NA</span></code></pre></div>
<p>Cut out the extra columns and spread the additive-vs-dominance across columns, just to tidy up</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">mtpred</span><span class="op">$</span><span class="va">varcovars</span> <span class="op">%&gt;%</span> 
  <span class="fu">tidyr</span><span class="fu">::</span><span class="fu"><a href="https://tidyr.tidyverse.org/reference/hoist.html">unnest_wider</a></span><span class="op">(</span><span class="va">varcomps</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span><span class="op">(</span><span class="op">-</span><span class="va">totcomputetime</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu">tidyr</span><span class="fu">::</span><span class="fu"><a href="https://tidyr.tidyverse.org/reference/nest.html">unnest</a></span><span class="op">(</span><span class="va">predictedfamvars</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu">tidyr</span><span class="fu">::</span><span class="fu"><a href="https://tidyr.tidyverse.org/reference/nest.html">unnest</a></span><span class="op">(</span><span class="va">predVars</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span><span class="op">(</span><span class="op">-</span><span class="va">PMV</span>,<span class="op">-</span><span class="va">Nsegsnps</span>,<span class="op">-</span><span class="va">totcomputetime</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu">spread</span><span class="op">(</span><span class="va">VarComp</span>,<span class="va">VPM</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="fu">rmarkdown</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/rmarkdown/man/paged_table.html">paged_table</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<div data-pagedtable="false">
<script data-pagedtable-source type="application/json">
{"columns":[{"label":["Trait1"],"name":[1],"type":["chr"],"align":["left"]},{"label":["Trait2"],"name":[2],"type":["chr"],"align":["left"]},{"label":["sireID"],"name":[3],"type":["chr"],"align":["left"]},{"label":["damID"],"name":[4],"type":["chr"],"align":["left"]},{"label":["VarA"],"name":[5],"type":["dbl"],"align":["right"]},{"label":["VarD"],"name":[6],"type":["dbl"],"align":["right"]}],"data":[{"1":"Trait1","2":"Trait1","3":"1","4":"1","5":"0.61733377","6":"0.348851611"},{"1":"Trait1","2":"Trait1","3":"1","4":"11","5":"0.40969370","6":"0.266463800"},{"1":"Trait1","2":"Trait1","3":"1","4":"2","5":"0.55700600","6":"0.362521101"},{"1":"Trait1","2":"Trait1","3":"11","4":"11","5":"0.20205363","6":"0.349387171"},{"1":"Trait1","2":"Trait1","3":"2","4":"11","5":"0.34936593","6":"0.355197316"},{"1":"Trait1","2":"Trait1","3":"2","4":"2","5":"0.49667822","6":"0.496796157"},{"1":"Trait1","2":"Trait2","3":"1","4":"1","5":"-0.40476347","6":"0.015341677"},{"1":"Trait1","2":"Trait2","3":"1","4":"11","5":"-0.25655749","6":"-0.006701879"},{"1":"Trait1","2":"Trait2","3":"1","4":"2","5":"-0.21180827","6":"0.036415684"},{"1":"Trait1","2":"Trait2","3":"11","4":"11","5":"-0.10835150","6":"-0.112798947"},{"1":"Trait1","2":"Trait2","3":"2","4":"11","5":"-0.06360228","6":"-0.015913548"},{"1":"Trait1","2":"Trait2","3":"2","4":"2","5":"-0.01885307","6":"-0.011286707"},{"1":"Trait2","2":"Trait2","3":"1","4":"1","5":"0.63675551","6":"0.253185007"},{"1":"Trait2","2":"Trait2","3":"1","4":"11","5":"0.50487233","6":"0.135295264"},{"1":"Trait2","2":"Trait2","3":"1","4":"2","5":"0.50189506","6":"0.241011072"},{"1":"Trait2","2":"Trait2","3":"11","4":"11","5":"0.37298914","6":"0.146313573"},{"1":"Trait2","2":"Trait2","3":"2","4":"11","5":"0.37001187","6":"0.213935073"},{"1":"Trait2","2":"Trait2","3":"2","4":"2","5":"0.36703460","6":"0.288984776"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
<p>Note that variance results <code>Trait1==Trait2</code> should match the single-trait results and co-variance predictions are now included.</p>
<p>Lastly, let’s quickly verify that the additive-only function works <em>and</em> that the <code>runMt___</code> functions won’t errory if we give just one cross / trait.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">mtpred_A</span><span class="op">&lt;-</span><span class="fu"><a href="../reference/runMtCrossVarPredsA.html">runMtCrossVarPredsA</a></span><span class="op">(</span>CrossesToPredict <span class="op">=</span> <span class="va">ped2predict</span>,
                              AlleleSubEffectList <span class="op">=</span> <span class="va">AddEffectsList</span>,
                              haploMat <span class="op">=</span> <span class="va">haploMat</span>,
                              recombFreqMat <span class="op">=</span> <span class="va">recombFreqMat</span>, 
                              ncores <span class="op">=</span> <span class="fl">6</span><span class="op">)</span>
<span class="co">#&gt; [1] "Variances predicted for family: 1x1- took 0.001 mins"</span>
<span class="co">#&gt; [1] "Variances predicted for family: 1x2- took 0.001 mins"</span>
<span class="co">#&gt; [1] "Variances predicted for family: 1x11- took 0.001 mins"</span>
<span class="co">#&gt; [1] "Variances predicted for family: 2x2- took 0.001 mins"</span>
<span class="co">#&gt; [1] "Variances predicted for family: 2x11- took 0.001 mins"</span>
<span class="co">#&gt; [1] "Variances predicted for family: 11x11- took 0.001 mins"</span>
<span class="co">#&gt; [1] "Done predicting fam vars. Took 0.05 mins for 6 families"</span>
<span class="co">#&gt; [1] "Variances predicted for family: 1x1- took 0.001 mins"</span>
<span class="co">#&gt; [1] "Variances predicted for family: 1x2- took 0.001 mins"</span>
<span class="co">#&gt; [1] "Variances predicted for family: 1x11- took 0 mins"</span>
<span class="co">#&gt; [1] "Variances predicted for family: 2x2- took 0.001 mins"</span>
<span class="co">#&gt; [1] "Variances predicted for family: 2x11- took 0.001 mins"</span>
<span class="co">#&gt; [1] "Variances predicted for family: 11x11- took 0.001 mins"</span>
<span class="co">#&gt; [1] "Done predicting fam vars. Took 0.05 mins for 6 families"</span>
<span class="co">#&gt; [1] "Variances predicted for family: 1x1- took 0 mins"</span>
<span class="co">#&gt; [1] "Variances predicted for family: 1x2- took 0 mins"</span>
<span class="co">#&gt; [1] "Variances predicted for family: 1x11- took 0.001 mins"</span>
<span class="co">#&gt; [1] "Variances predicted for family: 2x2- took 0.001 mins"</span>
<span class="co">#&gt; [1] "Variances predicted for family: 2x11- took 0.001 mins"</span>
<span class="co">#&gt; [1] "Variances predicted for family: 11x11- took 0 mins"</span>
<span class="co">#&gt; [1] "Done predicting fam vars. Took 0.05 mins for 6 families"</span></code></pre></div>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">mtpred_A</span><span class="op">$</span><span class="va">varcovars</span> <span class="op">%&gt;%</span> 
  <span class="fu">tidyr</span><span class="fu">::</span><span class="fu"><a href="https://tidyr.tidyverse.org/reference/hoist.html">unnest_wider</a></span><span class="op">(</span><span class="va">varcomps</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span><span class="op">(</span><span class="op">-</span><span class="va">totcomputetime</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu">tidyr</span><span class="fu">::</span><span class="fu"><a href="https://tidyr.tidyverse.org/reference/nest.html">unnest</a></span><span class="op">(</span><span class="va">predictedfamvars</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu">tidyr</span><span class="fu">::</span><span class="fu"><a href="https://tidyr.tidyverse.org/reference/nest.html">unnest</a></span><span class="op">(</span><span class="va">predVars</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu">rmarkdown</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/rmarkdown/man/paged_table.html">paged_table</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<div data-pagedtable="false">
<script data-pagedtable-source type="application/json">
{"columns":[{"label":["Trait1"],"name":[1],"type":["chr"],"align":["left"]},{"label":["Trait2"],"name":[2],"type":["chr"],"align":["left"]},{"label":["sireID"],"name":[3],"type":["chr"],"align":["left"]},{"label":["damID"],"name":[4],"type":["chr"],"align":["left"]},{"label":["VarComp"],"name":[5],"type":["chr"],"align":["left"]},{"label":["VPM"],"name":[6],"type":["dbl"],"align":["right"]},{"label":["PMV"],"name":[7],"type":["lgl"],"align":["right"]},{"label":["Nsegsnps"],"name":[8],"type":["int"],"align":["right"]},{"label":["totcomputetime"],"name":[9],"type":["dbl"],"align":["right"]}],"data":[{"1":"Trait1","2":"Trait1","3":"1","4":"1","5":"VarA","6":"0.61733377","7":"NA","8":"9","9":"0.031"},{"1":"Trait1","2":"Trait1","3":"1","4":"2","5":"VarA","6":"0.55700600","7":"NA","8":"18","9":"0.032"},{"1":"Trait1","2":"Trait1","3":"1","4":"11","5":"VarA","6":"0.40969370","7":"NA","8":"18","9":"0.034"},{"1":"Trait1","2":"Trait1","3":"2","4":"2","5":"VarA","6":"0.49667822","7":"NA","8":"12","9":"0.032"},{"1":"Trait1","2":"Trait1","3":"2","4":"11","5":"VarA","6":"0.34936593","7":"NA","8":"19","9":"0.033"},{"1":"Trait1","2":"Trait1","3":"11","4":"11","5":"VarA","6":"0.20205363","7":"NA","8":"11","9":"0.032"},{"1":"Trait2","2":"Trait2","3":"1","4":"1","5":"VarA","6":"0.63675551","7":"NA","8":"9","9":"0.031"},{"1":"Trait2","2":"Trait2","3":"1","4":"2","5":"VarA","6":"0.50189506","7":"NA","8":"18","9":"0.032"},{"1":"Trait2","2":"Trait2","3":"1","4":"11","5":"VarA","6":"0.50487233","7":"NA","8":"18","9":"0.028"},{"1":"Trait2","2":"Trait2","3":"2","4":"2","5":"VarA","6":"0.36703460","7":"NA","8":"12","9":"0.030"},{"1":"Trait2","2":"Trait2","3":"2","4":"11","5":"VarA","6":"0.37001187","7":"NA","8":"19","9":"0.033"},{"1":"Trait2","2":"Trait2","3":"11","4":"11","5":"VarA","6":"0.37298914","7":"NA","8":"11","9":"0.031"},{"1":"Trait1","2":"Trait2","3":"1","4":"1","5":"VarA","6":"-0.40476347","7":"NA","8":"9","9":"0.028"},{"1":"Trait1","2":"Trait2","3":"1","4":"2","5":"VarA","6":"-0.21180827","7":"NA","8":"18","9":"0.028"},{"1":"Trait1","2":"Trait2","3":"1","4":"11","5":"VarA","6":"-0.25655749","7":"NA","8":"18","9":"0.030"},{"1":"Trait1","2":"Trait2","3":"2","4":"2","5":"VarA","6":"-0.01885307","7":"NA","8":"12","9":"0.033"},{"1":"Trait1","2":"Trait2","3":"2","4":"11","5":"VarA","6":"-0.06360228","7":"NA","8":"19","9":"0.031"},{"1":"Trait1","2":"Trait2","3":"11","4":"11","5":"VarA","6":"-0.10835150","7":"NA","8":"11","9":"0.026"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">st_addefflist</span><span class="op">&lt;-</span><span class="va">AddEffectsList</span>
<span class="va">st_addefflist</span><span class="op">$</span><span class="va">Trait2</span><span class="op">&lt;-</span><span class="cn">NULL</span>
<span class="va">stpred_sc</span><span class="op">&lt;-</span><span class="fu"><a href="../reference/runMtCrossVarPredsA.html">runMtCrossVarPredsA</a></span><span class="op">(</span>CrossesToPredict <span class="op">=</span> <span class="va">ped2predict</span><span class="op">[</span><span class="fl">1</span>,<span class="op">]</span>,
                               AlleleSubEffectList <span class="op">=</span> <span class="va">st_addefflist</span>,
                               haploMat <span class="op">=</span> <span class="va">haploMat</span>,
                               recombFreqMat <span class="op">=</span> <span class="va">recombFreqMat</span>,
                               ncores <span class="op">=</span> <span class="fl">6</span><span class="op">)</span>
<span class="co">#&gt; [1] "Variances predicted for family: 1x1- took 0.001 mins"</span>
<span class="co">#&gt; [1] "Done predicting fam vars. Took 0.04 mins for 1 families"</span>
<span class="va">stpred_sc</span><span class="op">$</span><span class="va">varcovars</span> <span class="op">%&gt;%</span> 
  <span class="fu">tidyr</span><span class="fu">::</span><span class="fu"><a href="https://tidyr.tidyverse.org/reference/hoist.html">unnest_wider</a></span><span class="op">(</span><span class="va">varcomps</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span><span class="op">(</span><span class="op">-</span><span class="va">totcomputetime</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu">tidyr</span><span class="fu">::</span><span class="fu"><a href="https://tidyr.tidyverse.org/reference/nest.html">unnest</a></span><span class="op">(</span><span class="va">predictedfamvars</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu">tidyr</span><span class="fu">::</span><span class="fu"><a href="https://tidyr.tidyverse.org/reference/nest.html">unnest</a></span><span class="op">(</span><span class="va">predVars</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu">rmarkdown</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/rmarkdown/man/paged_table.html">paged_table</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<div data-pagedtable="false">
<script data-pagedtable-source type="application/json">
{"columns":[{"label":["Trait1"],"name":[1],"type":["chr"],"align":["left"]},{"label":["Trait2"],"name":[2],"type":["chr"],"align":["left"]},{"label":["sireID"],"name":[3],"type":["chr"],"align":["left"]},{"label":["damID"],"name":[4],"type":["chr"],"align":["left"]},{"label":["VarComp"],"name":[5],"type":["chr"],"align":["left"]},{"label":["VPM"],"name":[6],"type":["dbl"],"align":["right"]},{"label":["PMV"],"name":[7],"type":["lgl"],"align":["right"]},{"label":["Nsegsnps"],"name":[8],"type":["int"],"align":["right"]},{"label":["totcomputetime"],"name":[9],"type":["dbl"],"align":["right"]}],"data":[{"1":"Trait1","2":"Trait1","3":"1","4":"1","5":"VarA","6":"0.6173338","7":"NA","8":"9","9":"0.031","_row":"elapsed"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
</div>
<div id="a-note-on-marker-effects-input" class="section level1">
<h1 class="hasAnchor">
<a href="#a-note-on-marker-effects-input" class="anchor"></a>A note on marker effects input</h1>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="st">"AddEffectsList"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/str.html">str</a></span><span class="op">(</span><span class="va">AddEffectsList</span><span class="op">)</span>
<span class="co">#&gt; List of 2</span>
<span class="co">#&gt;  $ Trait1: num [1, 1:20] 0.218 0.03 0.232 0.343 -0.114 ...</span>
<span class="co">#&gt;   ..- attr(*, "dimnames")=List of 2</span>
<span class="co">#&gt;   .. ..$ : NULL</span>
<span class="co">#&gt;   .. ..$ : chr [1:20] "1_5" "1_6" "1_14" "1_16" ...</span>
<span class="co">#&gt;  $ Trait2: num [1, 1:20] -0.159 -0.297 -0.299 0.237 -0.716 ...</span>
<span class="co">#&gt;   ..- attr(*, "dimnames")=List of 2</span>
<span class="co">#&gt;   .. ..$ : NULL</span>
<span class="co">#&gt;   .. ..$ : chr [1:20] "1_5" "1_6" "1_14" "1_16" ...</span></code></pre></div>
<ul>
<li><p>List of allele substitution effects.</p></li>
<li><p>Each list element is a matrix. One matrix per trait.</p></li>
<li><p>Each element of the list is named with a string identifying the trait and the <code>colnames</code> of each matrix are labeled with snpIDs.</p></li>
<li>
<p>Dimensions of each matrix of marker effects</p>
<ul>
<li>
<span class="math inline">\(1 \times N_{snp}\)</span>: Many users will have marker effects estimated with REML or only want to use the posterior mean marker effects from a Bayesian marker-regression. For these users, the argument <code>predType=="VPM"</code> should be used <em>and</em> each matrix of the <code>AlleleSubEffectList</code> will have only one row.</li>
<li>
<span class="math inline">\(\frac{N_{iter} - N_{burnIn}}{N_{thin}} \times N_{snp}\)</span>: When predicting the Posterior Mean Variance (PMV), use argument <code>predType=="PMV"</code> when computing the predicted variance across a sample of marker effects, e.g. the thinned MCMC samples, usually stored on disk.</li>
</ul>
</li>
</ul>
</div>
<div id="references" class="section level1 unnumbered">
<h1 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h1>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-lehermeier2017" class="csl-entry">
Lehermeier, C., G. de los Campos, V. Wimmer, and C.-C. Schön. 2017. <span>“Genomic Variance Estimates: With or Without Disequilibrium Covariances?”</span> <em>Journal of Animal Breeding and Genetics</em> 134 (3): 232–41. <a href="https://doi.org/10.1111/jbg.12268">https://doi.org/10.1111/jbg.12268</a>.
</div>
<div id="ref-lehermeier2017a" class="csl-entry">
Lehermeier, Christina, Simon Teyssèdre, and Chris-Carolin Schön. 2017. <span>“Genetic Gain Increases by Applying the Usefulness Criterion with Improved Variance Prediction in Selection of Crosses.”</span> <em>Genetics</em>, October, genetics.300403.2017. <a href="https://doi.org/10.1534/genetics.117.300403">https://doi.org/10.1534/genetics.117.300403</a>.
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Marnin Wolfe.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
