<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Get started predicting genetic variances • predCrossVar</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Get started predicting genetic variances">
<meta property="og:description" content="predCrossVar">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">predCrossVar</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/predictCrossVars.html">Get started predicting genetic variances</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/wolfemd/predCrossVar/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="predictCrossVars_files/header-attrs-2.5/header-attrs.js"></script><link href="predictCrossVars_files/anchor-sections-1.0/anchor-sections.css" rel="stylesheet">
<script src="predictCrossVars_files/anchor-sections-1.0/anchor-sections.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Get started predicting genetic variances</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/wolfemd/predCrossVar/blob/master/vignettes/predictCrossVars.Rmd"><code>vignettes/predictCrossVars.Rmd</code></a></small>
      <div class="hidden name"><code>predictCrossVars.Rmd</code></div>

    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/wolfemd/predCrossVar">predCrossVar</a></span><span class="op">)</span></code></pre></div>
<div id="quick-simulation-of-input-data" class="section level1">
<h1 class="hasAnchor">
<a href="#quick-simulation-of-input-data" class="anchor"></a>Quick simulation of input data</h1>
<p>Use <strong>AlphaSimR</strong> to create quick simulation data to demonstrate the <strong>predCrossVar</strong> package.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://tidyverse.tidyverse.org">tidyverse</a></span><span class="op">)</span>; <span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://magrittr.tidyverse.org">magrittr</a></span><span class="op">)</span>;
<span class="co">#&gt; ── Attaching packages ─────────────────────────────────────── tidyverse 1.3.0 ──</span>
<span class="co">#&gt; ✓ ggplot2 3.3.2     ✓ purrr   0.3.4</span>
<span class="co">#&gt; ✓ tibble  3.0.4     ✓ dplyr   1.0.2</span>
<span class="co">#&gt; ✓ tidyr   1.1.2     ✓ stringr 1.4.0</span>
<span class="co">#&gt; ✓ readr   1.4.0     ✓ forcats 0.5.0</span>
<span class="co">#&gt; ── Conflicts ────────────────────────────────────────── tidyverse_conflicts() ──</span>
<span class="co">#&gt; x dplyr::filter() masks stats::filter()</span>
<span class="co">#&gt; x dplyr::lag()    masks stats::lag()</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Attaching package: 'magrittr'</span>
<span class="co">#&gt; The following object is masked from 'package:purrr':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     set_names</span>
<span class="co">#&gt; The following object is masked from 'package:tidyr':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     extract</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/gaynorr/AlphaSimR">AlphaSimR</a></span><span class="op">)</span>
<span class="co">#&gt; Loading required package: R6</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Attaching package: 'AlphaSimR'</span>
<span class="co">#&gt; The following object is masked from 'package:dplyr':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     mutate</span>
<span class="va">founderPop</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/AlphaSimR/man/quickHaplo.html">quickHaplo</a></span><span class="op">(</span>nInd<span class="op">=</span><span class="fl">100</span>, nChr<span class="op">=</span><span class="fl">2</span>, segSites<span class="op">=</span><span class="fl">50</span><span class="op">)</span>
<span class="va">baseSP</span><span class="op">&lt;-</span><span class="va"><a href="https://rdrr.io/pkg/AlphaSimR/man/SimParam.html">SimParam</a></span><span class="op">$</span><span class="fu">new</span><span class="op">(</span><span class="va">founderPop</span><span class="op">)</span>
<span class="va">baseSP</span><span class="op">$</span><span class="fu">addTraitAD</span><span class="op">(</span>nQtlPerChr <span class="op">=</span> <span class="fl">10</span>, 
                  mean <span class="op">=</span><span class="fl">10</span>, 
                  var <span class="op">=</span> <span class="fl">50</span>,
                  varD <span class="op">=</span> <span class="fl">20</span><span class="op">)</span>
<span class="va">pop</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/AlphaSimR/man/newPop.html">newPop</a></span><span class="op">(</span><span class="va">founderPop</span>,simParam <span class="op">=</span> <span class="va">baseSP</span><span class="op">)</span> <span class="co"># initial generation</span>
<span class="va">pop</span> <span class="op">%&lt;&gt;%</span> <span class="fu"><a href="https://rdrr.io/pkg/AlphaSimR/man/setPheno.html">setPheno</a></span><span class="op">(</span>simParam <span class="op">=</span> <span class="va">baseSP</span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># centimorgan scale map as a named vector</span>
<span class="co">## will be converted to recomb. freq. matrix in tutorial</span>
<span class="va">genmap</span><span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/pkg/AlphaSimR/man/getQtlMap.html">getQtlMap</a></span><span class="op">(</span>trait <span class="op">=</span> <span class="fl">1</span>,simParam <span class="op">=</span> <span class="va">baseSP</span><span class="op">)</span>
<span class="va">m</span><span class="op">&lt;-</span><span class="va">genmap</span><span class="op">$</span><span class="va">pos</span>; 
<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">m</span><span class="op">)</span><span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="va">genmap</span><span class="op">$</span><span class="va">chr</span>,<span class="st">"_"</span>,<span class="va">genmap</span><span class="op">$</span><span class="va">site</span><span class="op">)</span>
<span class="va">genmap</span><span class="op">&lt;-</span><span class="va">m</span>; <span class="fu"><a href="https://rdrr.io/r/base/rm.html">rm</a></span><span class="op">(</span><span class="va">m</span><span class="op">)</span>
<span class="va">haplos</span><span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/pkg/AlphaSimR/man/pullQtlHaplo.html">pullQtlHaplo</a></span><span class="op">(</span><span class="va">pop</span>,trait <span class="op">=</span> <span class="fl">1</span>,simParam <span class="op">=</span> <span class="va">baseSP</span><span class="op">)</span>
<span class="co"># for now, it seems the critical internal functions `predOneCrossVarA()` and `predOneCrossVarAD()` </span>
<span class="co"># require haplotypes for the same individual to be distinguished by "_HapA" and "_HapB"</span>
<span class="co"># will fix in a future version</span>
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">haplos</span><span class="op">)</span> <span class="op">%&lt;&gt;%</span> 
  <span class="fu"><a href="https://rdrr.io/r/base/grep.html">gsub</a></span><span class="op">(</span><span class="st">"_1"</span>,<span class="st">"_HapA"</span>,<span class="va">.</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="https://rdrr.io/r/base/grep.html">gsub</a></span><span class="op">(</span><span class="st">"_2"</span>,<span class="st">"_HapB"</span>,<span class="va">.</span><span class="op">)</span>
<span class="co"># column IDs for the loci should match in order and ID between all files</span>
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">haplos</span><span class="op">)</span><span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">genmap</span><span class="op">)</span>

<span class="co"># effects</span>
<span class="va">addEffects</span><span class="op">&lt;-</span><span class="va">baseSP</span><span class="op">$</span><span class="va">traits</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">@</span><span class="va">addEff</span>
<span class="va">domEffects</span><span class="op">&lt;-</span><span class="va">baseSP</span><span class="op">$</span><span class="va">traits</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">@</span><span class="va">domEff</span>
<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">addEffects</span><span class="op">)</span><span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">domEffects</span><span class="op">)</span><span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">haplos</span><span class="op">)</span></code></pre></div>
</div>
<div id="core-functions" class="section level1">
<h1 class="hasAnchor">
<a href="#core-functions" class="anchor"></a>Core functions</h1>
<p>At this early development stage, package is unforgiving. There are not many arguments with defaults. Input formats are pretty specific, but well defined.</p>
<p>The package helps automate prediction for both simple (one trait, one cross) and complex (multi-trait, many crosses) scenarios.</p>
<p><strong><code><a href="../reference/predOneCrossVarA.html">predOneCrossVarA()</a></code>:</strong> predict one cross, one variance (or covariance), additive-only model <code><a href="../reference/predCrossVarsA.html">predCrossVarsA()</a></code>: predict multiple crosses (potentially in parallel/multicore), one variance (or covariance), additive-only model <code><a href="../reference/runMtCrossVarPredsA.html">runMtCrossVarPredsA()</a></code>: wraps around <code><a href="../reference/predCrossVarsA.html">predCrossVarsA()</a></code>. Predicts the variances predict the variances <em>and</em> covariances for a set of crosses in terms of BVs (breeding values), <em>i.e.</em> additive-only model.</p>
<p>Equivalent functions for an additive-plus-dominance models are: <code><a href="../reference/predOneCrossVarAD.html">predOneCrossVarAD()</a></code> –&gt; <code><a href="../reference/predCrossVarsAD.html">predCrossVarsAD()</a></code> –&gt; <code><a href="../reference/runMtCrossVarPredsAD.html">runMtCrossVarPredsAD()</a></code>. I didn’t make a function to do only dominance…</p>
<p>There are also functions for the much less computationally challenging prediction of cross means.</p>
<p>The multi-trait, multi-family function, <code><a href="../reference/runMtCrossVarPredsA.html">runMtCrossVarPredsA()</a></code> should be suitable for the univariate, unifamily prediction <em>but</em> users might find the list-of-matrices input format requirement unnecessary.</p>
</div>
<div id="predict-genetic-variances---one-cross-one-trait" class="section level1">
<h1 class="hasAnchor">
<a href="#predict-genetic-variances---one-cross-one-trait" class="anchor"></a>Predict genetic variances - one cross, one trait</h1>
<p>To predict cross variance, 3 types of input are required:</p>
<ol style="list-style-type: decimal">
<li>Phased Parental Haplotypes</li>
<li>Genetic Map (or pairwise-recombination frequencies)</li>
<li>Marker Effects</li>
</ol>
<div id="genetic-map" class="section level2">
<h2 class="hasAnchor">
<a href="#genetic-map" class="anchor"></a>Genetic map</h2>
<p>The usual starting place for recombination frequency is a centimorgan-scale genetic map.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># utility function to compute matrix of pairwise recombination frequencies. </span>
<span class="va">recombFreqMat</span><span class="op">&lt;-</span><span class="fu"><a href="../reference/genmap2recombfreq.html">genmap2recombfreq</a></span><span class="op">(</span><span class="va">genmap</span>,nChr <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></code></pre></div>
<p>In practice, these matrices get big as they have dimension <span class="math inline">\(N_{snp} \times N_{snp}\)</span>. It is recommended to precompute and store the <code>recombFreqMat</code>.</p>
<p>In predicting genetic variance in the <span class="math inline">\(F_1\)</span> cross, the <code>recombFreqMat</code> the matrix <span class="math inline">\(1-2*recombFreqMat\)</span> is computed. To avoid having to do that separately for each family being predicted, I designed the prediction functions to require <span class="math inline">\(1-2*recombFreqMat\)</span> as input, to save a great deal of computation. <em>Therefore</em> pre-compute it and store <em>that</em> on disk. Probably this can be improved on!</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">recombFreqMat</span><span class="op">&lt;-</span><span class="fl">1</span><span class="op">-</span><span class="op">(</span><span class="fl">2</span><span class="op">*</span><span class="va">recombFreqMat</span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">recombFreqMat</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span>
<span class="co">#&gt;            1_6       1_7      1_11      1_13      1_18</span>
<span class="co">#&gt; 1_6  1.0000000 0.9600054 0.8153958 0.7514773 0.6127514</span>
<span class="co">#&gt; 1_7  0.9600054 1.0000000 0.8493658 0.7827844 0.6382791</span>
<span class="co">#&gt; 1_11 0.8153958 0.8493658 1.0000000 0.9216104 0.7514773</span>
<span class="co">#&gt; 1_13 0.7514773 0.7827844 0.9216104 1.0000000 0.8153958</span>
<span class="co">#&gt; 1_18 0.6127514 0.6382791 0.7514773 0.8153958 1.0000000</span></code></pre></div>
</div>
<div id="phased-parental-haplotypes" class="section level2">
<h2 class="hasAnchor">
<a href="#phased-parental-haplotypes" class="anchor"></a>Phased parental haplotypes</h2>
<p>The example dataset <strong>haplos</strong> contains 2 chrom. with 10 loci each (columns) for 100 individuals (200 rows).</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span><span class="op">(</span><span class="va">haplos</span><span class="op">)</span>
<span class="co">#&gt; [1] 200  20</span></code></pre></div>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">haplos</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span>
<span class="co">#&gt;        1_6 1_7 1_11 1_13 1_18</span>
<span class="co">#&gt; 1_HapA   1   1    0    0    1</span>
<span class="co">#&gt; 1_HapB   1   1    0    1    1</span>
<span class="co">#&gt; 2_HapA   0   0    1    0    1</span>
<span class="co">#&gt; 2_HapB   0   0    0    0    0</span>
<span class="co">#&gt; 3_HapA   0   1    0    0    0</span></code></pre></div>
</div>
<div id="marker-effects" class="section level2">
<h2 class="hasAnchor">
<a href="#marker-effects" class="anchor"></a>Marker effects</h2>
<p>For an additive-only model, the argument to <code><a href="../reference/runMtCrossVarPredsA.html">runMtCrossVarPredsA()</a></code> <strong>AlleleSubEffectList</strong></p>
<p>List of ALLELE SUBSTITUTION EFFECT matrices, one matrix per trait, Each element of the list is named with a string identifying the trait and the <code>colnames</code> of each matrix are labeled with snpIDs. Some (most?) users will have marker effects estimated with REML or only want to use the posterior mean marker effects from a Bayesian marker-regression. That is, some users will only For these (most) users, the argument <code>predType=="VPM"</code> should be used <em>and</em> each matrix of the <code>AlleleSubEffectList</code> will have only one row. Otherwise each matrix has <code><a href="https://rdrr.io/r/base/nrow.html">nrow((nIter-burnIn)/thin)</a></code> and <code><a href="https://rdrr.io/r/base/nrow.html">ncol(length(snpIDs))</a></code>. If only one trait is analyzed, the list <code>AlleleSubEffectList</code> can have just one named-element.</p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Marnin Wolfe.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
